### PR TITLE
Fix all ASAN issues in vectorscan

### PR DESCRIPTION
**Pull request by danlark1**
Closes https://github.com/VectorCamp/vectorscan/issues/91
Closes https://github.com/VectorCamp/vectorscan/issues/100

There were lots of problems in small ranges, I tried to fix them all by carefully reading the code.

Also adopt all changes we made to make it compile like setbit unsetting in tests.

We tested with all combinations HS_OPTIMIZE/DEBUG/HASWELL/SSE4.2/ARM/ASAN/MSAN/NOSANITIZERS. Haven't tested AVX512, PPC and SVE2

Resubmitting to a develop branch. I'll take a look at AVX512 failure from FAT_RUNTIME

### CODE REVIEW

**Comment by danlark1** [comment]
`-DSANITIZE=undefined`
`-DSANITIZE=address`
`-DSANITIZE=memory`

to cmake command

Some might not work with the ifunc attribute

**Comment by danlark1** [comment]
I believe I fixed all tests :)

**Comment by danlark1** [comment]
Overall feedback: I might guess I missed in a couple places some mask lowering because of zero byte inputs, however, in general, it took a while lot to comprehend and to match what was missing

Can you share what is the idea behind such major rewrites? I thought Vectorscan just replaced SIMD to portable ones and was very surprised to see ASAN issues and then I realized these major string search algos were rewritten.

Given hyperscan might be supported in the future, it makes new integrations much more complex and even if hyperscan decides not to support ARM for other reasons, merges will become exponentially complex. I saw just a couple optimizations which are missed in hyperscan and presented here. In the end we will have different bugs in two repos and from our view it makes just life more complex

Anyway, thanks for your work on portability, no issues with this, at least

**Comment by markos** [comment]
well, the idea with the refactoring is that original hyperscan is heavily tailored against Intel SIMD, using movemasks which are extremely expensive to emulate. My initial port was indeed just a SIMD port, but performance on Arm suffered heavily. After a lot of searching we decided we care only about API/ABI compatibility and would rather get better performance on other platforms even if it makes it easier to integrate code back into Vectorscan. In the unlikely event that Hyperscan decides to become more favourable against other architectures in the future, we will still support Vectorscan as I have found multiple places that performance would increase and code size reduced. I will just note that code size in the refactored places was reduced to almost 1/3rd of the original, and keeping more or less the same performance on Intel, but increasing performance on our Arm systems by almost 200% in some microbenchmarks. Furthermore, SIMD engines like SVE2 require much heavier refactorings than even the current because of the major breakthroughs. The original code, where all the loop functions are implemented for each SIMD engine just does not scale. 
Also, by being free from the limitations of the original project we are able to only keep relevant code -eg. Windows support was completely removed as unneeded and irrelevant. I'm actually pondering whether AVX512 support is actually beneficial at all, it makes the code much more complicated and so far the benefit in all but a few benchmarks has been negligible. But that's a decision for future. 
In short, expect more and heavier refactoring, however with the goal to always maintain API/ABI compatibility.

**Comment by markos** [comment]
Having said that, I should note that I will be adding ASAN to the CI, so such bugs should be caught early on next time. I do honestly appreciate the pointer for that and the PR of course.

**Comment by danlark1** [comment]
Thanks

I have also some fuzzing targets, I will try to find all differences between hyperscan and Vectorscan in terms of output
